### PR TITLE
OpenXR: change bindings to 'flatten' source paths

### PR DESCRIPTION
**Pull request by BastiaanOlij**
Our OpenXR action map currently stores multiple input/output paths (now called source paths) per action in an interaction profile.

This causes an issue with upcoming binding modifiers (see #97140).

This PR will "flatten" this structure so one binding records a single binding between an action and a source path. 
If multiple paths should be bound, multiple binding entries are needed.

This is technically a breaking change in that the action map will be altered in such a way that it can't be loaded by older versions of Godot, but it is required to implement future modifiers.

*Contributed by Khronos Group through the Godot Integration Project*

### CODE REVIEW

File: modules/openxr/action_map/openxr_interaction_profile.cpp [hunk 1]
```diff
@@ -65,38 +69,60 @@ Ref<OpenXRAction> OpenXRIPBinding::get_action() const {
...
+	return source_path == p_path;
 }
 
-void OpenXRIPBinding::add_path(const String p_path) {
-	if (!paths.has(p_path)) {
-		paths.push_back(p_path);
+void OpenXRIPBinding::add_path(const String p_path) { // deprecated
+	// Fallback logic, only assign first time this is called.
+	if (source_path != p_path) {
+		ERR_FAIL_COND(!source_path.is_empty());
```
Comment (dsnopek) [inline modules/openxr/action_map/openxr_interaction_profile.cpp:115]: This should probably have a human-readable message (via `ERR_FAIL_COND_MSG()`) because if a developer is using `add_path()` in their code for multiple paths, that code will need to be re-written, and so some clue would be good.

File: modules/openxr/action_map/openxr_interaction_profile.cpp [hunk 2]
```diff
@@ -65,38 +69,60 @@ Ref<OpenXRAction> OpenXRIPBinding::get_action() const {
...
 		emit_changed();
 	}
 }
 
-void OpenXRIPBinding::remove_path(const String p_path) {
-	if (paths.has(p_path)) {
-		paths.erase(p_path);
+void OpenXRIPBinding::remove_path(const String p_path) { // deprecated
+	// Fallback logic, clear if this is our path.
+	if (source_path == p_path) {
```
Comment (dsnopek) [inline modules/openxr/action_map/openxr_interaction_profile.cpp:124]: An error message would be nice here for when `source_path != p_path` for the same reason, since existing code dealing with multiple paths would need to be rewritten.

**Review by dsnopek — APPROVED** [review]
Overall, this restructuring of the action map data makes sense to me!

This is also a breaking change with regard to code that programmatically creates an action map with multiple paths per binding. However, I think it's pretty rare to programmatically create an action map, and even so, it would only affect bindings with multiple paths, which even further reduces the affected code. I personally think this is acceptable.

In 99.9% of cases, developers will create action maps in the editor, and this should correctly load and convert existing action maps (although, I haven't tested it - only reviewed the code).

TL;DR LGTM :-)

Only a handful of minor notes.

File: modules/openxr/action_map/openxr_interaction_profile.cpp [hunk 3]
```diff
@@ -151,32 +177,56 @@ Ref<OpenXRIPBinding> OpenXRInteractionProfile::get_binding(int p_index) const {
 }
 
 void OpenXRInteractionProfile::set_bindings(Array p_bindings) {
-	// TODO add check here that our bindings don't contain duplicate actions
+	bindings.clear();
+
+	for (int i = 0; i < p_bindings.size(); i++) {
+		Ref<OpenXRIPBinding> binding = p_bindings[i];
```
Comment (AThousandShips) [inline modules/openxr/action_map/openxr_interaction_profile.cpp:183]:
```suggestion
	for (Ref<OpenXRIPBinding> binding : p_bindings) {
```

File: modules/openxr/action_map/openxr_interaction_profile.cpp [hunk 4]
```diff
@@ -151,32 +177,56 @@ Ref<OpenXRIPBinding> OpenXRInteractionProfile::get_binding(int p_index) const {
...
 	}
 
 	return Ref<OpenXRIPBinding>();
 }
 
+Vector<Ref<OpenXRIPBinding>> OpenXRInteractionProfile::get_bindings_for_action(const Ref<OpenXRAction> p_action) const {
+	Vector<Ref<OpenXRIPBinding>> ret_bindings;
+
+	for (int i = 0; i < bindings.size(); i++) {
+		Ref<OpenXRIPBinding> binding = bindings[i];
```
Comment (AThousandShips) [inline modules/openxr/action_map/openxr_interaction_profile.cpp:216]:
```suggestion
	for (Ref<OpenXRIPBinding> binding : bindings) {
```

File: modules/openxr/action_map/openxr_interaction_profile.h [hunk 5]
```diff
@@ -41,26 +41,30 @@ class OpenXRIPBinding : public Resource {
 
 private:
 	Ref<OpenXRAction> action;
-	PackedStringArray paths;
+	String source_path;
+
+	// PackedStringArray paths;
```
[the remark this answers is not in the document]
Comment (BastiaanOlij) [reply modules/openxr/action_map/openxr_interaction_profile.h:46]: that was the planned, missed it :)

File: modules/openxr/action_map/openxr_interaction_profile.h [hunk 6]
```diff
@@ -41,26 +41,30 @@ class OpenXRIPBinding : public Resource {
...
+	static Ref<OpenXRIPBinding> new_binding(const Ref<OpenXRAction> p_action, const String &p_source_path); // Helper function for adding a new binding
 
 	void set_action(const Ref<OpenXRAction> p_action); // Set the action for this binding
 	Ref<OpenXRAction> get_action() const; // Get the action for this binding
 
-	int get_path_count() const; // Get the number of io paths
+	void set_source_path(const String &path);
+	String get_source_path() const;
+
+	// Deprecated
```
[the remark this answers is not in the document]
Comment (BastiaanOlij) [reply modules/openxr/action_map/openxr_interaction_profile.h:60]: We can't remove them all because `set_paths` and `get_paths` are still needed to convert old action maps to new. But yeah I can add that.

**Comment by BastiaanOlij** [comment]
I'm also renaming `source_path` to `binding_path` after an internal discussion that source path misleads developers in thinking the bindings are absolute. This is a common misperception around the action map where bindings are suggestions that the XR runtime can override either due to user preferences or due to other types of hardware being used.

File: modules/openxr/action_map/openxr_interaction_profile.cpp (hunk 3 above)
Comment (BastiaanOlij) [reply modules/openxr/action_map/openxr_interaction_profile.cpp:183]: Woop woop for iterators! :)

**Review by AThousandShips — APPROVED** [review]

File: modules/openxr/action_map/openxr_interaction_profile.cpp [hunk 7]
```diff
@@ -35,23 +35,29 @@ void OpenXRIPBinding::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_action"), &OpenXRIPBinding::get_action);
 	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "action", PROPERTY_HINT_RESOURCE_TYPE, "OpenXRAction"), "set_action", "get_action");
 
-	ClassDB::bind_method(D_METHOD("get_path_count"), &OpenXRIPBinding::get_path_count);
+	ClassDB::bind_method(D_METHOD("set_binding_path", "binding_path"), &OpenXRIPBinding::set_binding_path);
+	ClassDB::bind_method(D_METHOD("get_binding_path"), &OpenXRIPBinding::get_binding_path);
+	ADD_PROPERTY(PropertyInfo(Variant::STRING, "binding_path"), "set_binding_path", "get_binding_path");
+
+	// Deprecated
 	ClassDB::bind_method(D_METHOD("set_paths", "paths"), &OpenXRIPBinding::set_paths);
```
[the remark this answers is not in the document]
Comment (BastiaanOlij) [reply modules/openxr/action_map/openxr_interaction_profile.cpp:43]: I'd need to then also disable the code using these methods. It feels too impactful to me but I'm not really the target audience for running with deprecated code removed so I might be seeing a roadblock that isn't there.

File: modules/openxr/action_map/openxr_interaction_profile.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (dsnopek) [reply modules/openxr/action_map/openxr_interaction_profile.h]: I think if someone made a custom `deprecated=no` build, then it shouldn't be able to convert old action maps, only load those with the current format.

With a GDExtension that uses a deprecated method, loading it will actually fail in a `deprecated=no` build. I feel like failing to load old action maps is more or less equivalent?

File: modules/openxr/action_map/openxr_interaction_profile.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (dsnopek) [reply modules/openxr/action_map/openxr_interaction_profile.cpp:43]: As I wrote above, I also think loading old action maps should fail with a `deprecated=no` build.

**Review by dsnopek — APPROVED** [review]
Overall, this looks great to me!

I personally think we could go a little further with removing the code for old action maps in `deprecated=no` builds, but that shouldn't necessarily hold this up.

**Review by m4gr3d — APPROVED** [review]

File: modules/openxr/action_map/openxr_interaction_profile.h [hunk 8]
```diff
@@ -41,26 +41,29 @@ class OpenXRIPBinding : public Resource {
...
+	String get_binding_path() const;
 
-	void parse_paths(const String p_paths); // Parse a comma separated string of io paths.
-
-	bool has_path(const String p_path) const; // Has this io path
-	void add_path(const String p_path); // Add an io path
-	void remove_path(const String p_path); // Remove an io path
+	// Deprecated.
+	void set_paths(const PackedStringArray p_paths); // Set our paths (for loading from resource), needed for loading old action maps.
+	PackedStringArray get_paths() const; // Get our paths (for saving to resource), needed for converted old action maps.
```
[the remark this answers is not in the document]
Comment (AThousandShips) [reply modules/openxr/action_map/openxr_interaction_profile.h:60]: See discussion above

[the remark this answers is not in the document]
Comment (BastiaanOlij) [reply modules/openxr/action_map/openxr_interaction_profile.h:60]: As discussed above, I wasn't sure. The problem is that they are used/needed if someone hasn't upgraded their action map to the new format and may thus break someones application if they haven't specifically gone into the action map and re-saved their action map.

**Comment by BastiaanOlij** [comment]
Everything should be cleared up now, should be ready to be merged.

File: modules/openxr/doc_classes/OpenXRIPBinding.xml [hunk 9]
```diff
@@ -39,9 +39,13 @@
 	</methods>
 	<members>
 		<member name="action" type="OpenXRAction" setter="set_action" getter="get_action">
-			[OpenXRAction] that is bound to these paths.
+			[OpenXRAction] that is bound to [member binding_path].
 		</member>
-		<member name="paths" type="PackedStringArray" setter="set_paths" getter="get_paths" default="PackedStringArray()">
+		<member name="binding_path" type="String" setter="set_binding_path" getter="get_binding_path" default="&quot;&quot;">
+			Binding path that defines the input or output bound to [member action].
+			[b]Node:[/b] Binding paths are suggestions, an XR runtime may choose to bind the action to a different input or output emulating this input or output.
```
Comment (AThousandShips) [inline modules/openxr/doc_classes/OpenXRIPBinding.xml:46]:
```suggestion
			[b]Note:[/b] Binding paths are suggestions, an XR runtime may choose to bind the action to a different input or output emulating this input or output.
```

**Comment by Repiteo** [comment]
Thanks!
